lib/linked_list.rb

```diff
@@ -18,97 +18,195 @@ def initialize
+    # Space Complexity: O(1)
     def find_min
-      raise NotImplementedError
+      return nil if @head.nil?
+
+      min = @head.data
+      current = @head
+
+      until current.nil?
+        if current.data < min
+          min = current.data
+        end
+        current = current.next
+      end
+      return min
+
     end
 
 
     # Additional Exercises 
     # returns the value in the first node
     # returns nil if the list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1) The position of the head is always first, so no traversing needed to find value
+    # Space Complexity: O(1)
     def get_first
-      raise NotImplementedError
+      return nil if @head.nil?
+
+      return @head.data
     end
 
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n) adding to last so have to traverse list to the end first
+    # Space Complexity: O(1)
     def add_last(value)
-      raise NotImplementedError
+      # raise NotImplementedError
+      current = @head
+      if current.nil?
+        @head = Node.new(value)
+      else
+        current = current.next until current.next.nil?
+        current.next = Node.new(value)
+      end
     end
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def length
-      raise NotImplementedError
+      # raise NotImplementedError
+      current = @head
+      count = 0
+      until current.nil?
+        current = current.next
+        count += 1
+      end
+      return count
     end
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns nil if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_at_index(index)
-      raise NotImplementedError
+      return nil if @head.nil?
+
+      count = 0
+      current = @head
+      while count < index
+        current.next.nil? ? nil : current = current.next
+        count += 1
+      end
+      return current.data
     end
 
     # method to print all the values in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def visit
-      raise NotImplementedError
+      # raise NotImplementedError
+      return nil if @head.nil?
+
+      current = @head
+      until current.nil?
+        puts current.data
+        current = current.next
+      end
     end
 
     # method to delete the first node found with specified value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def delete(value)

```

Comment:
👍 Since you have to traverse the list until you find the node prior to the node you want to delete, the time complexity is O(n)